### DISCUSSION TITLE
What's the prescribed approach for performing a data lookup and redirect on the server side using cargo-leptos?

### DISCUSSION THREAD

**Discussion by saikatdas0790** [leptos-rs/leptos#2037] [Q&A] [answered]
My situation for routing looks someting like this:
- User lands on the root domain "/"
- Server performs a lookup to find the post to show this user from the DB
- It comes back with a list of (publisher_id, post_id) to display
- User needs to get shown a scrolling feed of posts that are uniquely identified by the (publisher_id, post_id) combination

The approach that I am considering is:
- On navigation to the root domain, a server function fetches a post from the database
- Once fetched, the server function redirects to the newly constructed URL which looks something like `/feed/publisher_id/post_id`
- This URL has a corresponding component that will have logic to fetch and render the view that needs to be served

What I'm trying to confirm is if the redirect from the root page to the feed specific page happens on the server itself or does it take a round trip where the server sends a redirect request to the client browser and then the browser makes a subsequent request to the the feed URL and then the server sends generated content for the same?

Is it possible to perform the redirection on the server and return a rendered page of the feed on the initial navigation without a round trip from the browser?

I'm trying to create an experience with the fastest time to initial render and this round trip would hinder that. Is this the right way to think about this?

An additional question around redirect logic. The redirect logic [here](https://github.com/leptos-rs/leptos/blob/a5f73b441c079f9138102b3a7d8d4828f045448c/examples/session_auth_axum/src/auth.rs#L154-L181) looks like it will be called from the Axum router [here](https://github.com/leptos-rs/leptos/blob/a5f73b441c079f9138102b3a7d8d4828f045448c/examples/session_auth_axum/src/main.rs#L91).

In the above case, for the behaviour I want, where should I list the router handler for the redirect to happen on the root domain? What should be my router entry? I imagine I can't set up the above wanted behaviour in the Leptos Router because the Leptos Router needs something that implements `IntoView` to be able to be listed as a route render output?

Thoughts?

**Comment by maartensson** [comment]
If all you need is to fetch a list of posts from the database and show it in a listview and detail view, then you don't need redirect.

Check this example: https://github.com/leptos-rs/leptos/blob/main/examples/ssr_modes/src/app.rs

**Reply by saikatdas0790** [reply]
No, but that's the thing, I want to do the redirect, but on the server side because the root domain is where all users will land and there will be some logic to determine the dynamic content to display to the user.

I would like that content to be fetched on the server and then to be rendered server side to deliver the rendered HTML/CSS/WASM to the client device while keeping the URL in sync so that it reflects the current navigation state of the app

**Reply by gbj** [reply]
I don't think you can literally redirect to a new URL on the server without a round trip to the browser — i.e., if you want to end up with a different URL in the browser address bar, you can't just magically do it from the server. In a hypothetical other world, the browser could request the URL `/` and the server could respond "actually, I'm giving you `/feed/publisher_id/post_id` with this HTML", but in the actual world I'm pretty sure it needs to go (browser) "give me `/`" (server) "got to `/feed/publisher_id/post_id` instead" (browser) "ok give me `/feed/publisher_id/post_id`" (server) "sure here's some HTML".

**Reply by saikatdas0790** [reply]
Okay. So, this means I need an additional round trip to the server. 

Isn't this possible?
Browser: Give me "/"
Server: Navigate/Redirect yourself to "/feed/publisher_id/post_id" and take this content and render it when navigated

**Reply by saikatdas0790** [reply]
And as a follow up question, I have logic for the redirect here:
https://github.com/go-bazzinga/hot-or-not-web-leptos-ssr/blob/ic-agent-integration/src/route/root/mod.rs

How do I get the Leptos router to pick this up? It has to be called on navigation to the root path

Since there's nothing to be rendered, there is no component code here

**Reply by gbj** [reply, marked as answer]
1. encapsulate that redirection logic in a server function
2. use `create_blocking_resource` to call that server function
3. read that resource in a Suspense

This may seem overly complicated but the benefit is that it behaves in exactly the same way on the client or the server: i.e., if someone initially loads `/`, it will serve them the redirect from the server. If they then navigate back to `/` on the client, it will call the server to get a new redirect URL and go there from the client.
